pkg/agent/controller/bgp/controller_test.go

```diff
@@ -0,0 +1,2043 @@
+		ingress := []corev1.LoadBalancerIngress{{IP: LoadBalancerIP}}
+		svc.Status.LoadBalancer.Ingress = ingress
+	}
+	return svc
+}
+
+func generateEgress(name string, ip string, nodeName string) *crdv1b1.Egress {
+	return &crdv1b1.Egress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			UID:  "test-uid",
+		},
+		Spec: crdv1b1.EgressSpec{
+			EgressIP: ip,
+		},
+		Status: crdv1b1.EgressStatus{
+			EgressIP:   ip,
+			EgressNode: nodeName,
+		},
+	}
+}
+
+func generateNode(name string, labels, annotations map[string]string) *corev1.Node {
+	return &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			UID:         "test-uid",
+			Labels:      labels,
+			Annotations: annotations,
+		},
+	}
+}
+
+func generateEndpointSlice(svcName string,
+	suffix string,
+	isLocal bool,
+	isIPv6 bool,
+	endpointIP string) *discovery.EndpointSlice {
+	addrType := discovery.AddressTypeIPv4
+	if isIPv6 {
+		addrType = discovery.AddressTypeIPv6
+	}
+	var nodeName *string
+	if isLocal {
+		nodeName = &localNodeName
+	}
+	protocol := corev1.ProtocolTCP
+	endpointSlice := &discovery.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", svcName, suffix),
+			Namespace: namespaceDefault,
+			UID:       "test-uid",
+			Labels: map[string]string{
+				discovery.LabelServiceName: svcName,
+			},
+		},
+		AddressType: addrType,
+	}
+	if endpointIP != "" {
+		endpointSlice.Endpoints = []discovery.Endpoint{{
+			Addresses: []string{
+				endpointIP,
+			},
+			Conditions: discovery.EndpointConditions{
+				Ready: ptr.To(true),
+			},
+			Hostname: nodeName,
+			NodeName: nodeName,
+		}}
+		endpointSlice.Ports = []discovery.EndpointPort{{
+			Name:     ptr.To("p80"),
+			Port:     ptr.To(int32(80)),
+			Protocol: &protocol,
+		}}
+	}
+	return endpointSlice
+}
+
+func generateBGPPeer(ip string, asn, port, gracefulRestartTimeSeconds int32) v1alpha1.BGPPeer {
+	return v1alpha1.BGPPeer{
+		Address:                    ip,
+		Port:                       &port,
+		ASN:                        asn,
+		MultihopTTL:                ptr.To(int32(1)),
+		GracefulRestartTimeSeconds: &gracefulRestartTimeSeconds,
+	}
+}
+
+func generateBGPPeerConfig(peerConfig *v1alpha1.BGPPeer, password string) bgp.PeerConfig {
+	return bgp.PeerConfig{
+		BGPPeer:  peerConfig,
+		Password: password,
+	}
+}
+
+func generateSecret(rawData map[string]string) *corev1.Secret {
+	data := make(map[string][]byte)
+	for k, v := range rawData {
+		data[k] = []byte(v)
+	}
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      types.BGPPolicySecretName,
+			Namespace: namespaceKubeSystem,
+			UID:       "test-uid",
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: data,
+	}
+}
+
+func ipStrToPrefix(ipStr string) string {
+	if utilnet.IsIPv4String(ipStr) {
+		return ipStr + ipv4Suffix
+	} else if utilnet.IsIPv6String(ipStr) {
+		return ipStr + ipv6Suffix
+	}
+	return ""
+}
+
+func waitAndGetDummyEvent(t *testing.T, c *fakeController) {
```

Comment:
nit: for a future PR, it may be good to rename this to `waitEvent` and `doneDummyEvent` could be renamed to `removeEvent`. They could also be defined as methods on `fakeController`.